File: LeetCode/python/copy_list_with_random_pointer.py
# A linked list is given such that each node contains an additional random pointer which could point to any node in the list or null.

# Return a deep copy of the list.

# Hide Tags [Hash Table] [Linked List]

# Definition for singly-linked list with a random pointer.
# class RandomListNode:
#     def __init__(self, x):
#         self.label = x
#         self.next = None
#         self.random = None

class Solution:
    # @param head, a RandomListNode
    # @return a RandomListNode
    def copyRandomList_1(self, head):
        if head == None: return
        # insert new node with same label value after each node
        tmp = head
        while tmp:
            newNode = RandomListNode(tmp.label)
            newNode.next, tmp.next = tmp.next, newNode
            tmp = tmp.next.next
        # copy random value
        tmp = head
        while tmp:
            if tmp.random:
                tmp.next.random = tmp.random.next
            tmp = tmp.next.next
        # seperate old and new linked list
        newhead = head.next
        pold, pnew = head, newhead
        while pnew.next:
            pold.next, pold = pnew.next, pold.next
            pnew.next, pnew = pold.next, pnew.next
        pold.next, pnew.next = None, None
        return newhead


    def copyRandomList_2(self, head):
        # Store mapping between old and new nodes
        mapOldToNew = {None:None}
        # Clone all old nodes with same label values
        curr = head
        while curr:
            mapOldToNew[curr] = RandomListNode(curr.label)
            curr = curr.next
        # Assign next and random value of cloned nodes
        curr = head
        while curr:
            mapOldToNew[curr].next = mapOldToNew[curr.next]
            mapOldToNew[curr].random = mapOldToNew[curr.random]
            curr = curr.next
        return mapOldToNew[head]

File: LintCode/java/CopyListWithRandomPointer.java
/**
 * Definition for singly-linked list with a random pointer.
 * class RandomListNode {
 *     int label;
 *     RandomListNode next, random;
 *     RandomListNode(int x) { this.label = x; }
 * };
 */
public class CopyListWithRandomPointer {
    /**
     * @param head: The head of linked list with a random pointer.
     * @return: A new head of a deep copy of the list.
     */

    // V1, O(n), O(n)
    // HashMap
    // Iterate twice
    public RandomListNode copyRandomList(RandomListNode head) {
        if (head == null) {
            return null;
        }

        HashMap<RandomListNode, RandomListNode> map = new HashMap<RandomListNode, RandomListNode>();
        RandomListNode dummy1 = new RandomListNode(0);
        RandomListNode dummy2 = new RandomListNode(0);
        RandomListNode head1 = head, head2;
        dummy1.next = head1;
        RandomListNode prev = dummy2;


        while (head1 != null) {
            head2 = new RandomListNode(head1.label);
            map.put(head1, head2);

            prev.next = head2;
            prev = prev.next;
            head1 = head1.next;
        }

        head1 = dummy1.next;
        head2 = dummy2.next;
        while (head1 != null) {
            head2.random = map.get(head1.random);
            head1 = head1.next;
            head2 = head2.next;
        }

        return dummy2.next;
    }

    // V2, O(n), O(n)
    // HashMap
    // Iterate once
    public RandomListNode copyRandomList(RandomListNode head) {
        if (head == null) {
            return null;
        }

        HashMap<RandomListNode, RandomListNode> map = new HashMap<RandomListNode, RandomListNode>();
        RandomListNode dummy = new RandomListNode(0);
        RandomListNode prev = dummy, newNode;

        while (head != null) {
            if (map.containsKey(head)) {
                newNode = map.get(head);
            } else {
                newNode = new RandomListNode(head.label);
                map.put(head, newNode);
            }

            prev.next = newNode;

            if (head.random != null) {
                if (map.containsKey(head.random)) {
                    newNode.random = map.get(head.random);
                } else {
                    newNode.random = new RandomListNode(head.random.label);
                    map.put(head.random, newNode.random);
                }
            }

            prev = prev.next;
            head = head.next;
        }

        return dummy.next;
    }

    // V3, O(n), O(1)
    // no hashmap/additional space version
    public RandomListNode copyRandomList(RandomListNode head) {
        if (head == null) {
            return null;
        }
        copyNext(head);
        copyRandom(head);
        return splitList(head);
    }
    private void copyNext(RandomListNode head) {
        while (head != null) {
            RandomListNode newNode = new RandomListNode(head.label);
            newNode.random = head.random;
            newNode.next = head.next;
            head.next = newNode;
            head = head.next.next;
        }
    }
    private void copyRandom(RandomListNode head) {
        while (head != null) {
            if (head.next.random != null) {
                head.next.random = head.random.next;
            }
            head = head.next.next;
        }
    }
    private RandomListNode splitList(RandomListNode head) {
        RandomListNode newHead = head.next;
        while (head != null) {
            RandomListNode tmp = head.next;
            head.next = tmp.next;
            head = head.next;
            if (tmp.next != null) {
                tmp.next = tmp.next.next;
            }
        }
        return newHead;
    }
}

